### PR TITLE
Add remaining indices and associated preprocessing code

### LINKED ISSUE #8: Do the work - develop script(s) to compute drought indices from ERA5 data and ACCAP climatologies
The main work for this repo/project. Develop scripts that download and process ERA5 data into indices useful for drought monitoring in Alaska, using climatologies developed by collaborators with ACCAP.

The following indices should be computed for 30, 60, 90, 180, and 365 days:

SPI - standardized precipitation index
SPEI (can ERA 5 do this?) – standardized potential evaporation index
Total Precip
% of Normal Precipitation
SWE - snow water equivalent
SWE % of Normal
Soil Moisture Deficit

### PR DESCRIPTION
**Pull request by kyleredilla**
This PR consists of changes for generating the penultimate dataset of drought indices. It includes the following high-level changes over the current `main` branch:

- swap of main processing code from jupyter notebook to python script - this will be more straightforward for automating/running
- code for generating the SPI, SPEI, and soil moisture deficit indices
- code used to download the precursor datasets needed for these indices: hourly soil moisture, precip, and potential evaporation data for the historical period
- code used to process the downloaded data into data inputs that will be utilized and provided with every execution of the total indices dataset generation (parameters of gamma distirbutions fit to precip and potential evaporation data for each day of the year)
- a quality control notebook used for doing point validations for all indices and intervals, plus histograms

To test, you really only need to follow the `README` to download the ERA5 data and run the pipeline, then use the QC notebook (`notebooks/snap/qc.ipynb`) to make sure the data appears correct. Use the `notebooks/explore_indices.ipynb` to further check that things seem to visualize OK (don't look too close though - the colormaps are not great). You really don't need to test all of the stuff in `scripts/snap`, just give it a look for understanding and typos. It shouldn't be needed again, or at least very often, but feel free to test if you'd like.

It is important that the `README` is meaningful for our collaborators that will be relying on it to compute the indices regularly, so keep a watchful eye for assumptions that might not make sense from an outsider's point of view. 

The data that will be provided to NWS should all be in the `/workspace/Shared/Tech_Projects/NWS_Drought_Indicators/project_data/nws_data`. You could work with this path, or copy it onto scratch space or your local machine for testing.

closes #8

### CODE REVIEW

File: scripts/snap/download_lsm.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+from config import (
+    api_credentials_check,
+    DL_BBOX,
+)
+
+api_credentials_check()
+
+
+def download_swvl_data():
+    """Download volumetric soil water layers 1 and 2 in yearly files for the climatology period of 1981-2020"""
```
Comment (charparr) [inline scripts/snap/download_lsm.py:19]:
```suggestion
    """Download the ERA5 land-sea mask file."""
```

File: scripts/snap/download_lsm.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import cdsapi
+from config import (
+    api_credentials_check,
+    DL_BBOX,
+)
+
+api_credentials_check()
+
+
+def download_swvl_data():
```
Comment (charparr) [inline scripts/snap/download_lsm.py:18]:
```suggestion
def download_land_sea_mask():
```

File: scripts/snap/download_lsm.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        },
+        lsm_fp,
+    )
+    
+    return
+
+    
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.INFO)
+    download_swvl_data()
```
Comment (charparr) [inline scripts/snap/download_lsm.py:44]:
```suggestion
   download_land_sea_mask()
```

File: scripts/snap/download_swvl.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+"""This is the script that was used to download the hourly ERA5 soil moisture data for computing a climatology. A combined version of the downloaded data will be provided alongside the other climatoglogies. This is provided here for reference and should not be needed by NWS/NOAA.
```
Comment (charparr) [inline scripts/snap/download_swvl.py:1]:
```suggestion
"""This is the script that was used to download the hourly ERA5 soil moisture data for computing a climatology. A combined version of the downloaded data will be provided alongside the other climatologies. This is provided here for reference and should not be needed by NWS/NOAA.
```

**Review by charparr — CHANGES_REQUESTED** [review]
I've left some comments and suggestions for the documentation of this pipeline - I've tried to take the position of someone who might not know much about this work - so if any of the comments or suggestions seem pedantic feel free to ignore them. I've successfully downloaded data and executed the drought indicators pipeline (`process.py`) for a January analysis date using the Conda environment provided here. I reviewed but did not run the scripts in `scripts/snap`- these produce the reference climatologies and parameters required by the pipeline, but are not something we expect external users to need to run.

I launched the `explore_indices` notebook. This notebook needs a bit more before I can get it running - I'm not sure how the input data is sourced and there are some commented out snippets of code that I'm unsure what to do with.

I also launched the `qc` notebook. I like the strategy here and this looks real thorough. I did hit an assertion error when checking the SPEI parameters:

```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
File <timed exec>:16

Cell In [4], line 100, in check_spei_params(daily_tp)
     95 with xr.open_dataset(CLIM_DIR.joinpath("spei_gamma_parameters.nc")) as spei_ds:
     96     test = spei_ds["params"].sel(**latlon_sel_di).sel(
     97         interval=interval, dayofyear=ref_doy
     98     )
--> 100 assert np.all(check == test)
    101 return spei_params

AssertionError: 
```
I'll take a closer look at this - I'm guessing it is a config or QC issue rather than a pipeline issue.

File: config.py [hunk 5]
```diff
@@ -3,20 +3,20 @@
...
+DOWNLOAD_DIR.mkdir(exist_ok=True)
 
 # final output dataset (NetCDF) of indices summarized over key intervals
-indices_fp = DOWNLOAD_DIR.joinpath("outputs/nws_drought_indices.nc")
-indices_fp.parent.mkdir(exist_ok=True)
+INDICES_DIR = DATA_DIR.joinpath("outputs")
+INDICES_DIR.mkdir(exist_ok=True)
 
 DATA_LAG_TIME_DAYS = 8
 DL_BBOX = [
```
[the remark this answers is not in the document]
Comment (kyleredilla) [reply config.py:24]: Good point, made a Trello card for now.

File: README.md [hunk 6]
```diff
@@ -1,5 +1,28 @@
 # NWS Drought Indicators
 
+This codebase was designed by SNAP for a collaboration with the National Weather Service and can be used for computing a series of "indicators" (aka indices) that may be useful in assessing drought conditions in Alaska. 
+
+It generates a dataset of 6 indicators computed over retrospective intervals from a supplied reference date with the following lengths (in days): 7, 30, 60, 90, 180, 365. Those indicators are:
+
+* `tp`: Total precipitation. $\sum p_i$ for all days $i$ in the summary interval.
+* `pntp`: Percent of normal total precipitation. $\frac{\sum p_j}{\sum pclim_j} * 100$ for all days-of-year $j$ in the summary interval, where $pclim_i$ is the climatological daily total precip value for day-of-year $i$.
```
[the remark this answers is not in the document]
Comment (kyleredilla) [reply README.md:8]: Yep I meant to change this one I think - end-of-the-day's work.

**Comment by kyleredilla** [comment]
Thanks for the very thorough review @charparr, great catches and fixes/ suggestions! 

I will try re-running from scratch to hopefully replicate that error in the QC notebook, and the issues with the interactive viz notebook.

**Review by charparr — APPROVED** [review]
I started from scratch and successfully downloaded the data and executed the pipeline to create the indicators. The QC notebook executed and all assertions passed with flying colors. The viz `explore_indices` notebooks is really strong - I love this and immediately started looking at all the different indicators and comparing across time chunks. So good.

**Comment by charparr** [comment]
Bruce is running a trial on a minimal VM (CentOS7 with Python 3.10.9 and Conda) and the download script is running fine at the moment. No need to hold up this PR and any further bumps found today/tonight will be ticketed!

**Pull request merged**